Propose update for simple output type expressions. (#40)
substrait-io/substrait: site/docs/expressions/scalar_functions.md

# Scalar Functions

> Scalar functions are a function that takes in values from a single record and produces an output value. To clearly specify the definition of functions, Substrait declares a extensible specification plus binding approach to function resolution. A scalar function signature includes the following properties:

Reviewer: ```suggestion
A function is a scalar function if that function takes in values from a single record and produces an output value. To clearly specify the definition of functions, Substrait declares a extensible specification plus binding approach to function resolution. A scalar function signature includes the following properties:
```
You could also maybe do "Scalar functions are functions that..."  but "Scalar functions are a function that" sounds weird to me.

This paragraph itself is rather intimidating and unclear to a neophyte but we tackle that more in the post-consensus phase.

| Property               | Description                                                  | Required                            |
| ---------------------- | ------------------------------------------------------------ | ----------------------------------- |

> | Name                   | One or more user friendly utf8 strings that are used to reference this function in languages | At least one value is required.     |

Reviewer: What does it mean if there are multiple names?  Are they aliases of each other?

Author: Yeah, they are aliases.

> | List of arguments      | Argument properties are defined below. Arguments can be fully defined, parameterized or wildcarded. See further details below. | Optional, defaults to niladic.      |

Reviewer: ```suggestion
| List of arguments      | Argument properties are defined below. Arguments can be fully defined or calculated with a type expression. See further details below. | Optional, defaults to niladic.      |
```

| Deterministic          | Whether this function is expected to reproduce the same output when it is invoked multiple times with the same input. This informs a plan consumer on whether it can constant reduce the defined function. An example would be a random() function, which is typically expected to be evaluated repeatedly despite having the same set of inputs. | Optional, defaults to true.         |

> | Session Dependent      | Whether this function is influenced by the session context it is invoked within. For example, a function may be influenced by a user who is invoking the function, the time zone of a session, or some other non-obvious parameter. This can inform caching systems on whether a particular function is cacheable. | Optional, defaults to false.        |

Reviewer: Naive question.  Do we really need both session dependent and deterministic or can session dependent just be one particular case of non-deterministic?  Will a system be able to optimize more efficiently knowing a function is deterministic but not session dependent?

Author: They play in different spaces/influence different optimization opportunities. 

Deterministic describes whether or not a value could be reduced to a constant in some expressions. If it is non-deterministic, you must execute for each record. Select rand(), a,b,c from foo is a good example. You wouldn't expect rand() to be the same value on any record. If you did the same thing as add(1,1), you could replace with the literal 2.

Session dependent describes whether or not a plan's interpretation is impacted by session properties and thus means you can't share something like a plan or data cache between different sessions (without also keying session properties). A function that returns username is a prime example.

> | Variadic Behavior      | Whether the last argument of the function is variadic. Options include: [Single argument, M..N]. M must be less than or equal to N. N can be defined or left as "unlimited". | Optional, defaults to single value. |

Reviewer: ```suggestion
| Variadic Behavior      | Whether the last argument of the function is variadic or a single argument.  If variadic the argument can optionally have a lower bound (minimum number of instances) and upper bound (maximum number of instances) | Optional, defaults to single value. |
```

`[Single argument, M..N]` <- I get what you are trying to express here but it doesn't come through to me except that I know the answer already.

Author: looks good

| Description            | Additional description of function for implementers or users. Should be written human readable to allow exposure to end users. Presented as a map with language => description mappings. E.g. `{ "en": "This adds two numbers together.", "fr": "cela ajoute deux nombres"}`. | Optional                            |

> | Return Type Expression | The output type is expected to be a physical type of the expression. Return types can be one of a simple, fully-defined compound type or a type expression. See below for more on type expressions. | Required                            |

Reviewer: ```suggestion
| Return Value | The output type of the expression.  Return types can be expressed as a fully-defined type or a type expression. See below for more on type expressions. | Required                            |
```

I don't know what you meant by "physical type".  I don't see the merit in distinguishing between a simple type and a fully-defined compound type.  They are both "fully-defined types".

Author: Good edit.

> | Implementation Map     | A map of implementation locations for one or more implementations of the given function. Each key is a function implementation type. Implementation types include examples such as: AthenaArrowLambda, TrinoV361Jar, ArrowCppKernelEnum, GandivaEnum, LinkedIn Transport Jar, etc. [Definition TBD]. Implementation type has one or more properties associated with retrieval of that implementation. | Optional                            |

Reviewer: Maybe a foolish ask but is there any chance we can get away with a URI here?

Author: My thinking was that providing the type of implementation in the map, it allows a system to only read manifests for implementations it can actually use. E.g. type => git-uri.

## Argument Types

There are two main types of arguments: Value Arguments and Type Arguments. 

> * Value Arguments: Arguments that where the function will refer to data value, could be a constant (literal-based expression) or variable (reference-based expression). This is the most common type of argument. Value arguments are not available in output derivation. Value arguments can be declared in one of two ways: concrete or parameterized. Concrete Types are either simple types or compound types with all parameters fully defined (without any parameters). Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc. Parameterized types are discussed further below.
> * Type Arguments: Type arguments are used to inform the evaluation and/or type derivation of the function. For example, you might have a function which is `truncate(<type> DECIMAL(P0,S0), <value> DECIMAL(P1, S1), <value> i32)`. This function declares two value arguments (arg1 and arg2) and type argument. The type argument's value can be used in output type determination. (The types of the values arguments can also be used for output type derivation.)

Reviewer: ```suggestion
* Value Arguments: An argument that refers to a data value.  This could be a constant (a literal expression defined in the plan) or variable (A reference expression that references data being processed by the plan). This is the most common type of argument. Value arguments are not available in output derivation. Value arguments can be declared in one of two ways: concrete or parameterized. Concrete Types are either simple types or compound types with all parameters fully defined (without referencing any type arguments). Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc. Parameterized types are discussed further below.
* Type Arguments: Type arguments are used to inform the evaluation and/or type derivation of the function. For example, you might have a function which is `truncate(<type> DECIMAL(P0,S0), <value> DECIMAL(P1, S1), <value> i32)`. This function declares two value arguments and a type argument. The type argument's value can be used to determine the output type (The types of the values arguments can also be used for output type derivation.)
```

I agree with what you are saying, I am just (hopefully) clearing up the wording.  `literal-based expression` and `reference-based expression` confused me for a while but I get what you are saying.  I removed `(arg1 and arg2)` because it isn't clear to me what you were trying to say.

Author: nice improvements, thanks

#### Value Argument Properties

| Property | Description                                                  | Required                                                   |
| -------- | ------------------------------------------------------------ | ---------------------------------------------------------- |

> | Name     | A human name for this argument to help clarify use.          | Optional, defaults to a name based on position (e.g. arg0) |

Reviewer: ```suggestion
| Name     | A human readable name for this argument to help clarify use.          | Optional, defaults to a name based on position (e.g. arg0) |
```

> | Type     | A fully concrete or parameterized (bound or not).            | Required                                                   |

Reviewer: ```suggestion
| Type     | A fully defined type or a type expression.            | Required                                                   |
```

| Constant | Whether this argument is required to be a constant for invocation. For example, in some system a regular expression pattern would only be accepted as a literal and not a column value reference. | Optional, defaults to false                                |

#### Type Argument Properties

| Property | Description                                                  | Required                                                   |
| -------- | ------------------------------------------------------------ | ---------------------------------------------------------- |
| Type     | A partially or completely parameterized type. E.g. List&lt;K&gt; or K | Required                                                   |

> | Name     | A human name for this argument to help clarify use.          | Optional, defaults to a name based on position (e.g. arg0) |

Reviewer: ```suggestion
| Name     | A human readable name for this argument to help clarify use.          | Optional, defaults to a name based on position (e.g. arg0) |
```

### Parameterized Types

> Types are parametered by two types of values: by inner types (e.g. List&lt;K&gt;) and numeric values (e.g. VARCHAR(P,S)). Parameter names are simple strings (frequently a single character). There are two types of parameters: integer parameters and type parameters. 

Reviewer: What about the names in a struct parameterization?  That doesn't seem to fit the dichotomy of types/integers.

Author: Yeah, there is no accomodation for that. For now, I would think that a plan builder would generate any required information using a type argument.

> Types are parametered by two types of values: by inner types (e.g. List&lt;K&gt;) and numeric values (e.g. VARCHAR(P,S)). Parameter names are simple strings (frequently a single character). There are two types of parameters: integer parameters and type parameters. 

Reviewer: ```suggestion
Types are parameterized by two types of values: by inner types (e.g. List&lt;K&gt;) and numeric values (e.g. VARCHAR(P,S)). Parameter names are simple strings (frequently a single character). There are two types of parameters: integer parameters and type parameters. 
```

Author: thanks

> When the same parameter name is used multiple times in a function definition, the function can only bind if the exact same value is used for all parameters of that name. For example, if one had a function with a signature of `fn(VARCHAR(N), VARCHAR(N)), the function would be only be usable if both VARCHAR types had the same length value N. This necessitates that all instances of the same parameter name must be of the same parameter type (all instances are a type parameter or always intances are aninteger parameter).

Reviewer: ```suggestion
When the same parameter name is used multiple times in a function definition, the function can only bind if the exact same value is used for all parameters of that name. For example, if one had a function with a signature of `fn(VARCHAR(N), VARCHAR(N)), the function would be only be usable if both VARCHAR types had the same length value N. This necessitates that all instances of the same parameter name must be of the same parameter type (all instances are a type parameter or all instances are an integer parameter).
```

It also necessitates that all instances of the same parameter name are the exact same value which seems a stronger conclusion.

#### Integer Parameter Bounds

> A type which declares one or more integer parameters can also declare a range of legal values for a function. For example, one could declare a function `fn(VARCHAR(N)) where N IN [0..20]` to describe a function which only takes in a varchar types that are less than or equal to 20 characters in length.

Reviewer: ```suggestion
A type which declares one or more integer parameters can also declare a range of legal values for a function. For example, one could declare a function `fn(VARCHAR(N)) where N IN [0..20]` to describe a function which only takes in varchar types that are less than or equal to 20 characters in length.
```

#### Type Parameter Bounds

> A parameterized type can either be unbounded (any type is allowed) or be bounded. When bounded, the function can only be bound to arguments of the types bound to. For example, one could declare a function `fn(Map<K,V> map) where K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` which would allow the function to accept a map argument as long as the key type is one of STRING, VARCHAR or FIXEDCHAR. Note, when duplicate parameter names are used in disjunct bounds options, the names will use the same bounds. For example, in the above map function, if one were to declare a bound such as where N IN [0..20], this limit would apply to either N parameter.

Reviewer: If I have `fn(K, K) where K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` is `fn(STRING, VARCHAR(N))` a valid combination?

Author: No. My goal with the wording previously was that K must the exact same value in all locations. If you want to support the pattern of mixed arguments, you'd need to use separate parameter names.

Parameters: type, integer
Literals: type, integer
```

Fully defined with argument types:

* `type_parameter(string name) => type`
* `integer_parameter(string name) => integer`
* `not(boolean x) => boolean` 
* `and(boolean a, boolean b) => boolean` 
* `or(boolean a, boolean b) => boolean` 
* `multiply(integer a, integer b) => integer`
* `divide(integer a, integer b) => integer`
* `add(integer a, integer b) => integer`
* `subtract(integer a, integer b) => integer`
* `min(integer a, integer b) => integer`
* `max(integer a, integer b) => integer`
* `equal(integer a, integer b) => boolean`
* `greater_than(integer a, integer b) => boolean`
* `less_than(integer a, integer b) => boolean`
* `covers(Type a, Type b) => boolean`Covers means that type b matches type A for as much as type B is defined. For example, if type A is `VARCHAR(20)` and type B is `VARCHAR(N)`, type B would be considered covering. Similarlily if type A was `List<Struct<a:f32, b:f32>>`and type B was `List<Struct<>>`, it would be considered covering. Note that this is directional "as in B covers A" or "B can be further enhanced to match the definition A". 
* `if(boolean a) then (integer) else (integer)`

> * `if(boolean a) then (type) else (type)`

Reviewer: I don't understand the purpose of this section.

Author: It declares the argument inputs/outputs for the available operations. The lines 77-80 were to make it easy to get a gist. The lines 83-101 were trying to be full definitions. (That was the intention, anyway.)

> #### Example Type Expressions Uses

Reviewer: ```suggestion
#### Example Type Expressions
```

Or maybe "Example Uses for Type Expressions" or "Example Use Cases for Type Expressions"

For reference, here are are some common output type derivations and how they can be expressed with a return type expression:

| Operation                                                 | Definition                                                   |
| --------------------------------------------------------- | ------------------------------------------------------------ |
| Add item to list                                          | `add(<List<T>, T>) => List<T>`                               |
| Decimal Division                                          | `divide(Decimal(P1,S1), Decimal(P2,S2)) => Decimal(P1 -S1 + S2 + MAX(6, S1 + P2 + 1), MAX(6, S1 + P2 + 1))` |

> | Do regex on only string maps to return values.            | `extract_values(Map<K,V>) => List<V> WHERE K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` |

Reviewer: ```suggestion
| Select a subset of map keys based on a regular expression (requires stringlike keys)           | `extract_values(Map<K,V>) => List<V> WHERE K IN [STRING, VARCHAR(N), FIXEDCHAR(N)]` |
```

Where is the regular expression expressed?

Author: hehehe, good point. Will add post accepting your edits.

> | Concatenate two fixed sized character strings             | `concat(FIXEDCHAR(A), FIXEDCHAR(B)) => CHAR(A+B)`            |

Reviewer: ```suggestion
| Concatenate two fixed sized character strings             | `concat(FIXEDCHAR(A), FIXEDCHAR(B)) => FIXEDCHAR(A+B)`            |
```

> | Make a struct of a set of fields and a struct definition. | `make_struct(<type> T, K...) => T`                           |

Reviewer: Where are the field names expressed?

Author: In the type T

Reviewer 2: What is the meaning of `K...` here? Does it mean every `K[0]` up to `K[n - 1]` are possibly distinct types? Or does it mean that every argument must be the same type `K`?

Also, what is the `<type> T` syntax? Isn't that covered by declaring type parameters in the function signature like `make_struct<T, K>(T, K...)`?

Author: >What is the meaning of K... here?
I called this out elsewhere in the doc. In a variadic case, the last argument declaration can be either marked as consistent or inconsistent (I need to add this to the protobuf). If consistent, must resolve the same. If inconsistent, each instance can be a different value.

> Also, what is the <type> T syntax?

It's all pseudo syntax to help explain the concepts for the spec. We need to define an actual syntax.

## Argument Types & Output Type Derivation

### Input Argument Types

Input arguments can be declared in one of two ways: materialized or parameterized.

* Materialized Type: Materialized types are either simple types or compound types with all parameters fully defined (without any parameters). Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc.
* Parameterized types: Types can be parameterized such that the parameter can be used in a output derivation e.g. `f(K) => K` or to guarantee consistency of type between input arguments `f(K,K) => boolean`. Example parameterized types would be VARCHAR(T), List&lt;E&gt;, MAP&lt;K, fp32&gt;, etc. A parameters is named with a simple UTF8 character or string.

### Direct Return Types

A direct return type is one that is fully known at function definition type. Example simple direct return types would be things such as i32, fp32. For compound types, a direct return type must be fully declared. Example fully defined types: VARCHAR(20), DECIMAL(25,5)

### Return Type Expressions

> Any function can declare a return type expression. A return type expression uses a simplified set of expressions to describe how the return type should be returned. For example, a return expression could be as simple as the return of parameter declared in the arguments. For example `f(List<K>) =>K ` or can be a simple mathematical or conditional expression such as `add(decimal(a,b), decimal(c,d)) => decimal(a+c, b+d)`. For the simple expression language, there is a very narrow set of types:

Author: In my model (probably not expressed well in the doc yet) there are two things: Type Parameters and Integer Parameters. Each have a name that can be referenced in output derivation if desired. I added TypeParameter to the list of available types in the type union to allow the expression List<K>. In the type definitions, integer parameters also replaced the direct integers in the types that take one or more integer parameters. As of right now, those are the only two types of parameters that types can be parameterized by.

* Integer: 64 bit signed integer (can be a literal or a parameter value)
* Boolean: True and False
* Type: A Substrait type (with possibly additional embedded expressions)

These types are evaluated using a small set of operations to support common scenarios. Since it is possible that output derivation ultimately determines that binding is not allowed, a `not_bindable` operation is provided that states that the particular binding is not allowed. List of valid operations:

```
Math: +, -, *, /, min, max
Boolean: &&, ||, !, <, >, ==
Input Parameters: type parameter, literal parameter
Literals: integer, type
Special: !bindable
```

Fully defined with argument types:

* `not(boolean x) => boolean` 
* `and(boolean a, boolean b) => boolean` 
* `or(boolean a, boolean b) => boolean` 
* `multiply(integer a, integer b) => integer`
* `divide(integer a, integer b) => integer`
* `add(integer a, integer b) => integer`

> * `substract(integer a, integer b) => integer`

Reviewer 2: I'm going to be that guy and suggest that we don't leave `integer` ambiguous here and give it an unambiguous integer type. I'd vote for `i64` since it's the widest type, supporting the most use cases. Hopefully this isn't an area of the spec where we need to be extremely concerned about integer sizes :)

Reviewer 2: Also, `substract` has an extra `s` in it.

Author: I guess I could make some kind of witty joke that anything sub* in the substrait project must be subst* and that is why I mispelled... :D   .fixed now.

Wrt to i64, I agree and actually already specified that just above.

* `min(integer a, integer b) => integer`
* `max(integer a, integer b) => integer`
* `equal(integer a, integer b) => boolean`
* `greater_than(integer a, integer b) => boolean`
* `less_than(integer a, integer b) => boolean`
* `equal(Type a, Type b) => boolean`
* `if(boolean a) then (integer) else (integer)`
* `if(boolean a) then (type) else (type)`

> * `not_bindable()`

Reviewer 2: It looks like there's `!bindable` and `not_bindable()`. Are these the same thing?

Reviewer 2: Oh, is `not_bindable()` what is produced by the syntax `!bindable`?

Author: None of this is syntax, just concepts. Syntax would be delegated to the serialization formats. In the abbreviated section I try to make it easy to understand what the ops are. The section below is blown out to better describe the parameters, etc.

#### Example Type Expressions Uses

For reference, here are are some common output type derivations and how they can be expressed with a return type expression:

| Operation                                     | Definition                                                   |
| --------------------------------------------- | ------------------------------------------------------------ |
| Add item to list                              | `add(<List<T>, T>) => List<T>`                               |

> | Decimal Division                              | `divide(Decimal(P1,S1), Decimal(P2,S2)) => Decimal(P1 -S1 + S2 + MAX(6, S1 + P2 + 1), MAX(6, S1 + P2 + 1))` |

Author: This is the common way these are expressed in SQL land. Brackets for type parameters. Parenthesis for integer parameters.

## Argument Types & Output Type Derivation

### Input Argument Types

Input arguments can be declared in one of three ways: complete, wildcarded, parameterized.

* Complete Type: Complete types are either simple types or compound types with all parameters fully defined. Examples include i32, fp32, VARCHAR(20), List&lt;fp32&gt;, etc.

> * Wildcard Types: Compound types can be wildcarded meaning that any valid value for the parameter can be accepted. Types are wildcarded by failing to express a particular value for the type. When serialized, this will typically mean using a null pointer or sentinel value. Examples would be  VARCHAR(*), List&lt;\*&gt;, DECIMAL(\*,10).

Reviewer 2: What is the difference between a parameterized type and a wildcard type in any of your examples? Is there a use case where a parameterized type doesn't cover the behavior of wildcards?

How is `List<*>` distinct from `List<T>`, for example?

Is the wildcard terminology standard outside of the database community? I haven't encountered this language before, but maybe I just haven't done enough reading.

Reviewer 2: I would expect for example a function like `list_length`, that operates on every cell of a list-of-any-type column to have a signature of `list_length<K>(list: List<K>) -> u32`.

Author: I distinguished them here to specifically state that a defined parameter name must be used in an expression to be as strict/clear as possible.

Author: I've updated to remove wildcards.

> * Parameterized types: Types can be parameterized such that the parameter can be used in a output derivation. Example parameterized types would be VARCHAR(T), List&lt;E&gt;, MAP&lt;K, fp32&gt;, etc. Parameters are simple names. Any parameters defined in an argument must be referenced at least once in an output expression. Parameterized argument types can only be used when a function is using expression based output type derivation.

Reviewer 2: > Any parameters defined in an argument must be referenced at least once in an output expression.

What is the rationale here?

Author: Per my comment above. A wildcard could also be called an unnamed parameter. A parameter is something that must be used. It is only to be as intentional as possible.

Reviewer 2: Hm, I guess I see this as analogous to an unused variable. It seems like introducing a new concept only to prevent people from not using type variables is pretty heavy weight since now someone has to build support for both, as well as the enforcement mechanism, when you can unify the notions of wildard and parameterized types under a single concept.

Author: Part of it was thinking about the serialization formats, which I guess we can ignore. When serializing something it seems weird to require a named unused parameter. If you say it is required at the spec level (e.g. you don't allow List<>, only List<K>) it seems like it would be confusing if the serialization then forced this on you. I suppose serialization could say either representation is fine and just leave it to implementors to have to tease out the distinction. In general, I'm trying to be more formal about these things as much as possible (and why I work on the protobuf sketch as I think about the problems).

I'm okay switching it but that was the reasoning :)

Reviewer: I agree with cpcloud here.  What about the case where it is unused in the output derivation, can be any type, but is not valid as a wildcard.  For example:

`EQUAL(FIXEDBINARY<*>, FIXEDBINARY<*>) -> boolean` is not valid because fixed binaries are only comparable if they have the same width.  However, the type parameters are not used in the output derivation.

Author: Good point on the equal. I've updated to remove the distinction between wildcard and parameterized.

### Direct Return Types

> A direct return type is one that is fully known at function definition type. Example simple direct return types would be things such as i32, fp32. For compound types, a direct return type must be fully declared. Example fully defined types: VARCHAR(20), DECIMAL(25,5)

Reviewer 2: > A direct return type is one that is fully known at function definition type.

Should the last `type` be `time`?

Also, I find this language confusing. A function definition with a parameterized type is fully known at function definition time. It won't necessarily have been monomorphized at definition time, is that what you're referring to here? Maybe I'm being a bit pedantic but my thinking is that we shouldn't invent new words for things :)

Author: Feel free to propose alternative language here.

Reviewer: What is function definition time?  The point when the function is written down in the spec?  Or the point when the query plan is created?

If the former, perhaps simply:

A direct return type is one that does not depend on any parameterized input type.

If the latter:

A direct return type is one that is not know until the query has begun executing.

I assume from the context that we are talking about the former.

Author: Yes, the former. I'll update in next patch.

### Return Type Expressions

Any function can declare a return type expression. A return type expression uses a simplified set of expressions to describe how the return type should be returned. For example, a return expression could be as simple as the return of parameter declared in the arguments. For example `f(List<K>) =>K ` or can be a simple mathematical or conditional expression such as `add(decimal(a,b), decimal(c,d)) => decimal(a+c, b+d)`. For the simple expression language, there is a very narrow set of types:

* Integer: 64 bit signed integer (can be a literal or a parameter value)
* Boolean: True and False
* Type: A Substrait type (with possibly additional embedded expressions)

> These types are evaluated using a small set of operations to to support common scenarios. Since it is possible that output derivation ultimately determines that binding is not allowed, a `not_bindable` operation is provided that states that the particular binding is now sllowed. List of valid operations:

Reviewer 2: > These types are evaluated using a small set of operations to to support common scenarios.

Double `to to`.

Reviewer 2: > binding is now sllowed

Should this be `binding is not allowed`?

Author: both fixed

```
Math: +, -, *, /, min, max
Boolean: &&, ||, !, <, >, ==
Input Parameters: type parameter, literal parameter
Literals: integer, type

> Special: !bindable

Reviewer 2: Should `!bindable` have some way of customizing the error message? Maybe not, just wondering.

Author: I wasn't thinking so. I wouldn't expect this information to be propogated to a user. I could imagine a situation where a user writes a query and 50 functions are initially evaluated but then found unbindable. Then on the 51, a binding works and is used. The fact that things can't bind feels like an internal plan producer concern and they should decide how to expose to users.

```

Fully defined with argument types:

* `not(boolean x) => boolean` 
* `and(boolean a, boolean b) => boolean` 
* `or(boolean a, boolean b) => boolean` 
* `multiply(integer a, integer b) => integer`
* `divide(integer a, integer b) => integer`
* `add(integer a, integer b) => integer`
* `substract(integer a, integer b) => integer`
* `min(integer a, integer b) => integer`
* `max(integer a, integer b) => integer`
* `equal(integer a, integer b) => boolean`
* `greater_than(integer a, integer b) => boolean`
* `less_than(integer a, integer b) => boolean`
* `equal(Type a, Type b) => boolean`
* `if(boolean a) then (integer) else (integer)`
* `if(boolean a) then (type) else (type)`
* `not_bindable()`

#### Example Type Expressions Uses

For reference, here are are some common output type derivations and how they can be expressed with a return type expression:

| Operation                                     | Definition                                                   |
| --------------------------------------------- | ------------------------------------------------------------ |

> | Add item to list                              | `add(<List<T>, T>) => List<T>)`                              |

Reviewer 2: You have an unmatched paren at the end here.

Author: fixed

| Decimal Division                              | `divide(Decimal(P1,S1), Decimal(P2,S2)) => Decimal(P1 -S1 + S2 + MAX(6, S1 + P2 + 1), MAX(6, S1 + P2 + 1))` |

> | Do regex on only string maps to return values | `extract_values(Map<K,V>) => if(K==STRING) THEN LIST<V> ELSE !bindable` |

Reviewer 2: Why wouldn't this just be written as `extract_values(Map<String, V>)`?

Author: Fair point, my original plan was to do multiple if statements. E.g. k==string or k==varchar. I just got lazy.

Author: Updated

Reviewer 2: Maybe I missed it but is the plan to keep this as simple as possible? So, for example, if someone wanted to write a derivation function that corresponded to "any non-compound type", they're in for a giant `or` expression.

Author: My goal is that the language should be as simple as reasonable. I'm open to adding a in-list type idea but I'm not entirely convinced it is necessary. All non-compound types doesn't actually seem like a real thing. Numeric types probably seems more real.

Reviewer 2: > All non-compound types doesn't actually seem like a real thing.

Well, I did base that example on extracting values from a map, and I think you'd probably at least want any non-compound type to be able to be part of a a map. :shrug: I don't feel too strongly about it, keeping it simple to start seems like a good idea.

Author: Hmm... not super important at this moment but I wonder if we're using compound types to mean different things. I meant compound types as done in substrait where it is any type that is parameterized (including things like decimal, varchar, etc). Using that definition, I think most people would probably want to treat string and varchar much the the same and could imagine both being used as a key to map.

I better understand your comment about what maps would be restricted to if you limit "compound" to be types that contain other types such as struct, map and list. I've generally been referring to that set as "complex" as opposed to "compound". (Happy to revise terminology if something else is better for people.)

Reviewer 2: Ah yes! Sorry. I was definitely referring to what you've called _complex_ types, not _compound_ types.

Does that change your response at all? Fine if not, just wondering if you think "all non-complex types" is a predicate that will be commonly desired and used.

Author: My general inclination would be to expose a generic OR_EQUALITY_LIST(value, type, type,...,type) operation as opposed to declare specific classes. Specific classes have the problem that if you then declared u8, for example, it wouldn't be in the "numeric" group. I'm torn on whether OR is enough or OR_EQUALITY_LIST is necessary. Love to hear your gut and others.

Reviewer 2: Before I answer that, are we expecting that every consumer/producer provides an output derivation where they are required, or will there be a default implementation? I could imagine not needing to worry about `OR` vs `IN` vs `OR_EQUALITY_LIST` (`OR` would probably suffice) if we have the right defaults to avoid having to repeat the same giant `OR` expression in every producer/consumer.

Author: The derivation definition would be in the function signature yaml (or whatever format/name we give it), not the actual plan. In a function signature yaml, I'd be happy for that format to support named type expressions to normalize that kind of pattern (so it doesn't have be declared repetitively).

Reviewer 2: I see. In that case I suspect `OR` would be fine to start. Seems like it'd be very clear pretty quickly if an `IN` style operator/function was needed.
